internal/selectors/selectors.go

```diff
@@ -15,7 +15,38 @@ import (
 	api "sigs.k8s.io/hierarchical-namespaces/api/v1alpha2"
 )
 
-func ShouldPropagate(inst *unstructured.Unstructured, nsLabels labels.Set) (bool, error) {
+func SelectorExists(inst *unstructured.Unstructured, nsLabels labels.Set) (bool, error) {
+	if sel, err := GetSelector(inst); err != nil {
+		return false, err
+	} else if sel != nil && !sel.Empty() {
+		return true, nil
+	}
+	if sel, err := GetTreeSelector(inst); err != nil {
+		return false, err
+	} else if sel != nil && !sel.Empty() {
+		return true, nil
+	}
+	if none, err := GetNoneSelector(inst); err != nil || none {
+		return true, err
+	}
+	if all, err := GetAllSelector(inst); err != nil || all {
+		return true, err
+	}
+
+	return false, nil
+}
+
+func ShouldPropagate(inst *unstructured.Unstructured, nsLabels labels.Set, propagateByDefault bool) (bool, error) {
+	if !propagateByDefault {
+		// if propagateByDefault is false and there are no selectors - do not propagate,
+		// otherwise handle propagation as if propagateByDefault mode is true
+		if exists, err := SelectorExists(inst, nsLabels); err != nil {
+			return false, err
+		} else if !exists {
+			return false, nil
+		}
+	}
+
```

Comment:
I think this can just be moved to the last line of the function. Instead of just `return true, nil` it should be `return propagateByDefault, nil`.

Although if we change the function signature to accept the propagation mode, it should be `return mode == api.Propagate, nil` with a comment about what the alternatives are to 'propagate'.

Reply:
Yes this is a lot cleaner and clearer, I agree.